examples/erc721/tests/erc721.rs

```diff
@@ -1352,6 +1352,234 @@ async fn is_approved_for_all_invalid_operator(
+    let data: Bytes = Bytes::new();
+
+    let receipt =
+        receipt!(contract.safeMint(receiver_address, token_id, data.clone()))?;
+
+    assert!(receipt.emits(Erc721::Transfer {
+        from: Address::ZERO,
+        to: receiver_address,
+        tokenId: token_id,
+    }));
+
+    assert!(receipt.emits(ERC721ReceiverMock::Received {
+        operator: alice.address(),
+        from: Address::ZERO,
+        tokenId: token_id,
+        data,
+    }));
+
+    let Erc721::ownerOfReturn { ownerOf } =
+        contract.ownerOf(token_id).call().await?;
+    assert_eq!(receiver_address, ownerOf);
+
+    Ok(())
+}
+
+#[e2e::test]
+async fn safe_mint_to_receiver_contract_with_data(
+    alice: Account,
+) -> eyre::Result<()> {
+    let contract_addr = alice.as_deployer().deploy().await?.address()?;
+    let contract = Erc721::new(contract_addr, &alice.wallet);
+
+    let receiver_address =
+        receiver::deploy(&alice.wallet, ERC721ReceiverMock::RevertType::None)
+            .await?;
+
+    let token_id = random_token_id();
+    let data: Bytes = fixed_bytes!("deadbeef").into();
+
+    let receipt =
+        receipt!(contract.safeMint(receiver_address, token_id, data.clone()))?;
+
+    assert!(receipt.emits(Erc721::Transfer {
+        from: Address::ZERO,
+        to: receiver_address,
+        tokenId: token_id,
+    }));
+
+    assert!(receipt.emits(ERC721ReceiverMock::Received {
+        operator: alice.address(),
+        from: Address::ZERO,
+        tokenId: token_id,
+        data,
+    }));
+
+    let Erc721::ownerOfReturn { ownerOf } =
+        contract.ownerOf(token_id).call().await?;
+    assert_eq!(receiver_address, ownerOf);
+
+    Ok(())
+}
+
+#[e2e::test]
+async fn error_when_safe_mint_to_contract_without_receiver_interface_with_data(
+    alice: Account,
+) -> eyre::Result<()> {
+    let contract_addr = alice.as_deployer().deploy().await?.address()?;
+    let contract = Erc721::new(contract_addr, &alice.wallet);
+
+    let token_id = random_token_id();
+    let data: Bytes = fixed_bytes!("deadbeef").into();
+
+    let err = send!(contract.safeMint(contract_addr, token_id, data.clone()))
+        .expect_err("should not transfer the token to invalid receiver");
+
+    assert!(err.reverted_with(Erc721::ERC721InvalidReceiver {
+        receiver: contract_addr
+    }));
+
+    Ok(())
+}
+
+#[e2e::test]
+async fn error_when_safe_mint_to_invalid_sender_with_data(
+    alice: Account,
+    bob: Account,
+) -> eyre::Result<()> {
+    let contract_addr = alice.as_deployer().deploy().await?.address()?;
+    let contract = Erc721::new(contract_addr, &alice.wallet);
+
+    let invalid_sender = Address::ZERO;
+    let token_id = random_token_id();
+
+    let data: Bytes = fixed_bytes!("deadbeef").into();
+    _ = watch!(contract.mint(alice.address(), token_id))?;
+
+    let err = send!(contract.safeMint(bob.address(), token_id, data.clone()))
+        .expect_err("should not mint with an invalid sender");
```

Comment:
```suggestion
        .expect_err("should not safe mint an existing token");
```
a more accurate reason for the error